src/main/scala/firrtl/passes/RemoveAccesses.scala

```diff
@@ -16,90 +16,81 @@ object RemoveAccesses extends Pass {
     */
-  case class Location(base: Expression, guard: Expression)
+  private case class Location(base: Expression, guard: Expression)
 
   /** Walks a referencing expression and returns a list of valid references
     * (base) and the corresponding guard which, if true, returns that base.
     * E.g. if called on a[i] where a: UInt[2], we would return:
     *   Seq(Location(a[0], UIntLiteral(0)), Location(a[1], UIntLiteral(1)))
     */
-  def getLocations(e: Expression): Seq[Location] = e match {
+  private def getLocations(e: Expression): Seq[Location] = e match {
     case e: WRef => create_exps(e).map(Location(_,one))
     case e: WSubIndex =>
       val ls = getLocations(e.exp)
       val start = get_point(e)
       val end = start + get_size(tpe(e))
       val stride = get_size(tpe(e.exp))
-      val lsx = mutable.ArrayBuffer[Location]()
-      for (i <- 0 until ls.size) {
-        if (((i % stride) >= start) & ((i % stride) < end)) {
-          lsx += ls(i)
-        }
-      }
-      lsx
+      for ((l, i) <- ls.zipWithIndex
+        if ((i % stride) >= start) & ((i % stride) < end)) yield l
     case e: WSubField =>
       val ls = getLocations(e.exp)
       val start = get_point(e)
       val end = start + get_size(tpe(e))
       val stride = get_size(tpe(e.exp))
-      val lsx = mutable.ArrayBuffer[Location]()
-      for (i <- 0 until ls.size) {
-        if (((i % stride) >= start) & ((i % stride) < end)) { lsx += ls(i) }
-      }
-      lsx
+      for ((l, i) <- ls.zipWithIndex
+        if ((i % stride) >= start) & ((i % stride) < end)) yield l
     case e: WSubAccess =>
       val ls = getLocations(e.exp)
       val stride = get_size(tpe(e))
       val wrap = tpe(e.exp).asInstanceOf[VectorType].size
-      val lsx = mutable.ArrayBuffer[Location]()
-      for (i <- 0 until ls.size) {
+      ls.zipWithIndex map {case (l, i) =>
         val c = (i / stride) % wrap
-        val basex = ls(i).base
-        val guardx = AND(ls(i).guard,EQV(uint(c),e.index))
-        lsx += Location(basex,guardx)
+        val basex = l.base
+        val guardx = AND(l.guard,EQV(uint(c),e.index))
+        Location(basex,guardx)
       }
-      lsx
   }
   /** Returns true if e contains a [[firrtl.WSubAccess]]
     */
-  def hasAccess(e: Expression): Boolean = {
-    var ret: Boolean = false
-    def rec_has_access(e: Expression): Expression = e match {
-      case (e:WSubAccess) => { ret = true; e }
-      case (e) => e map (rec_has_access)
+  private def hasAccess(e: Expression): Boolean = {
+    def rec_has_access(e: Expression): Boolean = e match {
+      case (e:WSubAccess) => true
+      case (e:WSubField) => rec_has_access(e.exp)
+      case (e:WSubIndex) => rec_has_access(e.exp)
+      case (e:Mux) => rec_has_access(e.cond) || rec_has_access(e.tval) || rec_has_access(e.fval)
+      case (e:ValidIf) => rec_has_access(e.cond) || rec_has_access(e.value)
+      case (e:DoPrim) => e.args exists rec_has_access
+      case _ => false
     }
```

Comment:
I saw some terrible stack overflows w.r.t mappers before, so I just removed it if I can do easily. I'm totally fine to revert this change if this doesn't cause any problems.